posts/schrodingers-pineapple-upside-down-cake-front-end-development-for-back-end-engineers.md

```diff
@@ -0,0 +1,371 @@
+
+### Execution Contexts and Function Binding
+
+We can fix the error caused by clicking on the no-information card's button by updating our event listener ever so slightly, from `buttonElement.addEventListener('click', that.printButtonClick);` to `buttonElement.addEventListener('click', that.printButtonClick.bind(that));`. 
+
+![Page and console output with binding in place, after pressing buttons on the no-information and Richmond cards.](../images/2022/02/fed4bees3-function-binding.png)
+
+Okay, fine. But _why_? 
+
+We'll start by demonstrating what's going on. We can update the `printButtonClick` to first log `this` to the console, and add both versions of the function (the regular one and the one bound to `that`) as event listeners. 
+
+![Page and console output demonstrating that `this` is the button element when not bound, whereas it is the whole card component element when bound.](../images/2022/02/fed4bees3-demonstrating-what-this-means.png)
+
+We see that within the unbound function, `this` refers to the `button` element itself, whereas within the bound function, `this` refers to the `weather-card-component` element, which was what we wanted. Indeed, it's more than what we wanted; that's what we expected based on experience with Java. In Java, a method within a class is an instance method (unless it's static), and calling the method on an instance means that `this` always refers to that instance. Instance methods couple together two slightly different concerns: where to find the method and what object to operate them on. 
+
+Analogy time! From just the perspective of syntax and scope, let's imagine a class as a building, and a method or function defined within the class as a room in the building ("the room where it happens," as they say). There's a frame on a wall of the room that will show you what `this` is. An instance of a Java class is like a suburban home with a family living in it. For the concern of finding the method, a neighbor can point you to the Smith house and likely can tell you where the living room is one you get inside. For the concern of knowing what object to operate on, we look at the frame on the wall and find a somewhat impressionist pastel rendering of the front of the Smiths' house. It was a gift from the builder, so each house in the neighborhood has a similar frame in the living room, depicting the house it's in. So calling an instance method in Java is like finding a map of the neighborhood, navigating to the right house, stepping into the correct room, looking at the frame on the wall to determine `this`, and then... doing something. What folks do in the privacy of their own homes is none of our business!
```

Comment:
👏🏼  Love this explanation of syntax and scope! It's such a unique way to describe the concepts!